manageiq-operator/pkg/helpers/miq-components/network_policies.go

```diff
@@ -0,0 +1,237 @@
+	networkPolicy := newNetworkPolicy(cr, "-allow-postgres")
+
+	f := func() error {
+		if err := controllerutil.SetControllerReference(cr, networkPolicy, scheme); err != nil {
+			return err
+		}
+		addAppLabel(cr.Spec.AppName, &networkPolicy.ObjectMeta)
+		setIngressPolicyType(networkPolicy)
+
+		networkPolicy.Spec.PodSelector.MatchLabels = map[string]string{"name": "postgresql"}
+
+		pod := orchestratorPod(*c)
+		if pod == nil {
+			return nil
+		}
+
+		ensureIngressRule(networkPolicy)
+		setFirstIngressTCPPort(networkPolicy, 5432)
+		if len(networkPolicy.Spec.Ingress[0].From) != 2 {
+			networkPolicy.Spec.Ingress[0].From = []extensionsv1beta1.NetworkPolicyPeer{
+				extensionsv1beta1.NetworkPolicyPeer{},
+				extensionsv1beta1.NetworkPolicyPeer{},
+			}
+		}
+		orchestratedByLabelKey := cr.Spec.AppName + "-orchestrated-by"
+		orchestratedByLabelValue := pod.Name
+		networkPolicy.Spec.Ingress[0].From[0].PodSelector = &metav1.LabelSelector{}
+		networkPolicy.Spec.Ingress[0].From[0].PodSelector.MatchLabels = map[string]string{"name": "orchestrator"}
+		networkPolicy.Spec.Ingress[0].From[1].PodSelector = &metav1.LabelSelector{}
+		networkPolicy.Spec.Ingress[0].From[1].PodSelector.MatchLabels = map[string]string{orchestratedByLabelKey: orchestratedByLabelValue}
+
+		return nil
+	}
+
+	return networkPolicy, f
+}
+
+func newNetworkPolicy(cr *miqv1alpha1.ManageIQ, name string) *extensionsv1beta1.NetworkPolicy {
+	return &extensionsv1beta1.NetworkPolicy{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      cr.Spec.AppName + name,
+			Namespace: cr.ObjectMeta.Namespace,
+		},
+	}
+}
+
+func orchestratorPod(c client.Client) *corev1.Pod {
+	podList := &corev1.PodList{}
+	c.List(context.TODO(), podList)
+
+	for _, pod := range podList.Items {
+		if pod.ObjectMeta.Labels["name"] == "orchestrator" {
+			return &pod
+		}
+	}
+
+	return nil
+}
+
+func setIngressPolicyType(networkPolicy *extensionsv1beta1.NetworkPolicy) {
+	if len(networkPolicy.Spec.PolicyTypes) != 1 {
+		networkPolicy.Spec.PolicyTypes = append(networkPolicy.Spec.PolicyTypes, "Ingress")
+	}
+	networkPolicy.Spec.PolicyTypes[0] = "Ingress"
+}
+
+func ensureIngressRule(networkPolicy *extensionsv1beta1.NetworkPolicy) {
+	if len(networkPolicy.Spec.Ingress) != 1 {
+		networkPolicy.Spec.Ingress = []extensionsv1beta1.NetworkPolicyIngressRule{
+			extensionsv1beta1.NetworkPolicyIngressRule{},
+		}
+	}
+}
+
+func setFirstIngressTCPPort(networkPolicy *extensionsv1beta1.NetworkPolicy, port int32) {
+	if len(networkPolicy.Spec.Ingress[0].Ports) != 1 {
+		networkPolicy.Spec.Ingress[0].Ports = []extensionsv1beta1.NetworkPolicyPort{
+			extensionsv1beta1.NetworkPolicyPort{},
+		}
+	}
+	tcp := corev1.ProtocolTCP
+	networkPolicy.Spec.Ingress[0].Ports[0].Protocol = &tcp
+	networkPolicy.Spec.Ingress[0].Ports[0].Port = &intstr.IntOrString{Type: intstr.Int, IntVal: port}
```

Comment:
Can we DRY up some of this code, with something like -

```go
ports := networkPolicy.Spec.Ingress[0].Ports
if len(ports) != 1 {
  ...

ports[0].Protocol = &tcp
...
```

Reply:
I don't think so.  On initial pass (network policy creation), the ports is then empty.  If we then added `networkPolicy.Spec.Ingress[0].Ports = ports` to the end it could work, but that typically causes problems in the reconcile loop sending unnecessary updates when nothing has changed.

Reply:
I don't follow.... I'm talking within the method...does `.Ports` change within the pass of the method?

Reply:
This method is called repeatedly, once on operator startup then every reconcile and its results are passed to `CreateOrUpdate`.

For the initial creation, if we added the setter `networkPolicy.Spec.Ingress[0].Ports = ports` that would work fine.  That will create the network policy as expected.  In the process, kubernetes will fill in any other default fields to define the object.

Next time reconcile is called, that would cause problems because that line will overwrite the definition of the ports (without any of the kubernetes defaults).  This will send a patch to the API, unnecessarily updating the NetworkPolicy definition.  Kubernetes will fill in the defaults again, and every time reconcile is called, this will happen again.

If we put a watch on the NetworkPolicies, this will put the operator into a never-ending reconcile .loop

Reply:
Is `networkPolicy.Spec.Ingress[0].Ports` a method-with-side-effects or a property?  I'm really lost, and I suspect it's because I don't understand Go.

Reply:
Golang student here too.

All of these usually access and update the kubernetes objects (json) using well defined structure definitions which otherwise would be accessing generic unstructured hash/maps.

Here you can follow the NetworkPolicy definitions down to the specific Ports being accessed/updated.



https://godoc.org/k8s.io/api/extensions/v1beta1#NetworkPolicy
https://godoc.org/k8s.io/api/extensions/v1beta1#NetworkPolicySpec
https://godoc.org/k8s.io/api/extensions/v1beta1#NetworkPolicyIngressRule
https://godoc.org/k8s.io/api/extensions/v1beta1#NetworkPolicyPort

Reply:
The problem for me isn't the structures, it's the code itself, or perhaps, what you can do with structs in the code...

I guess my question is, at a fundamental go level...what is the difference between:

```go
ports := networkPolicy.Spec.Ingress[0].Ports
ports[0].Protocol = &foo
ports[0].Port = &bar
```

and

```go
networkPolicy.Spec.Ingress[0].Ports[0].Protocol = &foo
networkPolicy.Spec.Ingress[0].Ports[0].Port = &bar
```

Reply:
I played around in the Go playground for a bit, and I think I understand the difference in that 

```go
ports := networkPolicy.Spec.Ingress[0].Ports
```

creates a _copy_ of the struct, so changing it does nothing to the final object.  However if you take a reference to it, then you can dry it up...

```go
ports := &networkPolicy.Spec.Ingress[0].Ports
ports[0].Protocol = &foo
ports[0].Port = &bar
```

Reply:
Done